Fix fails query if bucket number exceeds declared bucket size

ArvinZheng opened the pull request:
fix #4378

sopel39 commented on presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:
```diff
@@ -636,6 +639,23 @@ private static boolean shouldUseFileSplitsFromInputFormat(InputFormat<?, ?> inpu
+        // validate the bucket number detected from files, fail the query if the highest bucket number detected from file
+        // exceeds the allowed highest number
+        if (highestBucketNumber >= partitionBucketCount) {
+            throw new PrestoException(HIVE_INVALID_BUCKET_FILES, format(
```
nit: each argument in newline

ArvinZheng replied:
fixed

sopel39 commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java:
```diff
@@ -566,6 +569,41 @@ public void testHive2VersionedFullAcidTableFails()
+            BackgroundHiveSplitLoader.validateFileBuckets(bucketFiles, 20, "tableName", "partitionName");
+            BackgroundHiveSplitLoader.validateFileBuckets(bucketFiles, 30, "tableName", "partitionName");
+        }
+        catch (Throwable e) {
```
remove catch, throwing an exception will cause test to fail anyway

electrum replied:
These are expected to be positive test cases that don't throw. Having try/catch is strange, as any exception will cause a test failure. By this logic, we would need try/catch in every test, as we normally expect that a test will not throw.

ArvinZheng replied:
got it, will remove it

sopel39 commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java:
```diff
@@ -566,6 +570,41 @@ public void testHive2VersionedFullAcidTableFails()
+            BackgroundHiveSplitLoader.validateFileBuckets(bucketFiles, 30, "tableName", "partitionName");
+        }
+        catch (Throwable e) {
+            Assert.shouldNeverReachHere();
```
static import

ArvinZheng replied:
fixed

sopel39 commented on presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:
```diff
@@ -636,6 +639,28 @@ private static boolean shouldUseFileSplitsFromInputFormat(InputFormat<?, ?> inpu
+            return;
+        }
+
+        Integer highestBucketNumber = Collections.max(bucketFiles.keySet());
```
static import `max`

ArvinZheng replied:
fixed

sopel39 commented on presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:
```diff
@@ -636,6 +639,28 @@ private static boolean shouldUseFileSplitsFromInputFormat(InputFormat<?, ?> inpu
+        if (highestBucketNumber >= partitionBucketCount) {
+            throw new PrestoException(
+                    HIVE_INVALID_BUCKET_FILES,
+                    format("Hive table '%s' is corrupt. The highest bucket number in the directory (%s) is greater than the highest bucket " +
```
each arg in newline

ArvinZheng replied:
reformatted again

ArvinZheng commented:
@sopel39 thank you very much for your time.

electrum commented:
The commit message title is too long. Please shorten to something like

> Fail query if there are too many buckets

We follow these guidelines: https://chris.beams.io/posts/git-commit/

When updating the PR, please squash the commits into a single commit (we don't want to merge "fixup" commits).

electrum commented on presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:
```diff
@@ -636,6 +639,27 @@ private static boolean shouldUseFileSplitsFromInputFormat(InputFormat<?, ?> inpu
+            return;
+        }
+
+        Integer highestBucketNumber = max(bucketFiles.keySet());
```
This can be declared as `int`

ArvinZheng replied:
good catch

electrum commented on presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:
```diff
@@ -636,6 +639,27 @@ private static boolean shouldUseFileSplitsFromInputFormat(InputFormat<?, ?> inpu
+                            "declared (%s) for partition: %s",
+                    tableName,
+                    highestBucketNumber,
+                    partitionBucketCount - 1,
```
Remove `- 1` after changing error message

ArvinZheng replied:
I was thinking the same when I was trying to organize the error message, then I chose to make this trade off since I was not able to figure out a better one - how about `Hive table '%s' is corrupt. The highest bucket number in the directory (%s) exceeds the bucket number range defined by the declared bucket count (%s) for partition: %s`

ArvinZheng replied:
one error message example that may look a bit odd, feel free to suggest a better one 
> Hive table 'tableName' is corrupt. The highest bucket number in the directory (9) exceeds the bucket number range defined by the declared bucket count (9) for partition: partitionName

electrum replied:
I see what you mean. Maybe we could list the range explicitly?

> Hive table 'tableName' is corrupt. The highest bucket number in the directory (9) exceeds the bucket number range (0..8) defined by the declared bucket count (9) for partition: partitionName

electrum replied:
I merged the PR since this seems good enough for now. If you want to try adding the range or otherwise improving the error message, feel free to send a follow up PR.

ArvinZheng replied:
Yup, that's my concern. To show the bucket range I'll need to do calculation on the `partitionBucketCount`, maybe extra a method will have better readability for the code, what do you think?
BTW, thanks for your time on reviewing the code.

electrum replied:
The only calculation is `partitionBucketCount - 1` so I don't think we need a method (the range always starts at zero).

ArvinZheng replied:
Yes, my concerns is whether `subtracting 1` from `partitionBucketCount` is easy for people to understand, otherwise i am good to go with `partitionBucketCount - 1` directly.

electrum commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestBackgroundHiveSplitLoader.java:
```diff
@@ -566,6 +571,41 @@ public void testHive2VersionedFullAcidTableFails()
+        bucketFiles.put(6, null);
+        bucketFiles.put(9, null);
+
+        assertThatThrownBy(() -> BackgroundHiveSplitLoader.validateFileBuckets(bucketFiles, 1, "tableName", "partitionName"))
```
Static import `validateFileBuckets` for readability

electrum replied:
You can do this using
```java
assertPrestoExceptionThrownBy(() -> validateFileBuckets(bucketFiles, 1, "tableName", "partitionName"))
        .hasErrorCode(HIVE_INVALID_BUCKET_FILES)
        .hasMessage(...);

ArvinZheng replied:
got it

ArvinZheng commented:
@electrum For sure I will squash all the commits into one and shorten the commit message after all comments are addressed.
But what do you think about
> Fail query if there are invalid buckets

compare to

> Fail query if there are too many buckets

BTW, if you don't have any comments I'll go ahead and squash the commit once I get your response on the commit mesage.

electrum commented:
@ArvinZheng either one sounds good to me. Apart from the review comments, I don't have anything else. It should be ready to merge after those addressed. Thanks for the quick response.

ArvinZheng commented:
@electrum addressed all the comments except for the error message, replied you inline, trying to figure out a better sentence, mind taking a look, thanks.

electrum commented:
Thanks!
